File: scripts/upload-data.js
const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');

// Initialize Firebase Admin SDK
// Expects FIREBASE_SERVICE_ACCOUNT environment variable to contain the JSON key
const serviceAccount = JSON.parse(process.env.FIREBASE_SERVICE_ACCOUNT);

admin.initializeApp({
    credential: admin.credential.cert(serviceAccount)
});

const db = admin.firestore();

async function uploadCollection(directory, collectionName) {
    const dirPath = path.join(__dirname, '..', directory);

    if (!fs.existsSync(dirPath)) {
        console.warn(`Directory ${dirPath} does not exist. Skipping.`);
        return;
    }

    const files = fs.readdirSync(dirPath).filter(file => file.endsWith('.json'));

    console.log(`Uploading ${files.length} files from ${directory} to collection ${collectionName}...`);

    for (const file of files) {
        const filePath = path.join(dirPath, file);
        const fileContent = fs.readFileSync(filePath, 'utf8');
        const data = JSON.parse(fileContent);

        // Use filename without extension as document ID (e.g., 'draft_data_2023')
        const docId = path.basename(file, '.json');

        try {
            await db.collection(collectionName).doc(docId).set(data);
            console.log(`Successfully uploaded ${docId}`);
        } catch (error) {
            console.error(`Error uploading ${docId}:`, error);
        }
    }
}



async function calculateAndUploadStats() {
    console.log('Calculating all-time stats...');
    const fantasyDir = path.join(__dirname, '..', 'data', 'fantasy');
    if (!fs.existsSync(fantasyDir)) return;

    const files = fs.readdirSync(fantasyDir).filter(file => file.endsWith('.json'));

    let stats = {
        seasons_count: 0,
        total_games: 0,
        total_points: 0,
        highest_score: { value: 0, team: '', week: '', season: '' },
        lowest_score: { value: 1000, team: '', week: '', season: '' },
        largest_margin: { value: 0, winner: '', loser: '', week: '', season: '' },
        most_points_season: { value: 0, team: '', season: '' }
    };

    stats.seasons_count = files.length;

    for (const file of files) {
        const season = file.match(/fantasy_data_(\d+)\.json/)[1];
        const content = JSON.parse(fs.readFileSync(path.join(fantasyDir, file), 'utf8'));

        let seasonPoints = {};

        if (content.weeks) {
            Object.entries(content.weeks).forEach(([weekNum, weekData]) => {
                if (weekData.matchups) {
                    weekData.matchups.forEach(matchup => {
                        stats.total_games++;

                        // Process Team 1
                        if (matchup.team1) {
                            const score1 = parseFloat(matchup.team1.score || 0);
                            stats.total_points += score1;

                            // High Score
                            if (score1 > stats.highest_score.value) {
                                stats.highest_score = { value: score1, team: matchup.team1.name, week: weekNum, season: season };
                            }
                            // Low Score (ignore 0 if valid scores exist, but keeping logical check)
                            if (score1 > 0 && score1 < stats.lowest_score.value) {
                                stats.lowest_score = { value: score1, team: matchup.team1.name, week: weekNum, season: season };
                            }

                            // Season Points Accumulator
                            const t1Name = matchup.team1.name;
                            if (!seasonPoints[t1Name]) seasonPoints[t1Name] = 0;
                            seasonPoints[t1Name] += score1;
                        }

                        // Process Team 2
                        if (matchup.team2) {
                            const score2 = parseFloat(matchup.team2.score || 0);
                            stats.total_points += score2;

                            // High Score
                            if (score2 > stats.highest_score.value) {
                                stats.highest_score = { value: score2, team: matchup.team2.name, week: weekNum, season: season };
                            }
                            // Low Score
                            if (score2 > 0 && score2 < stats.lowest_score.value) {
                                stats.lowest_score = { value: score2, team: matchup.team2.name, week: weekNum, season: season };
                            }

                            // Season Points Accumulator
                            const t2Name = matchup.team2.name;
                            if (!seasonPoints[t2Name]) seasonPoints[t2Name] = 0;
                            seasonPoints[t2Name] += score2;
                        }

                        // Margin
                        if (matchup.team1 && matchup.team2) {
                            const score1 = parseFloat(matchup.team1.score || 0);
                            const score2 = parseFloat(matchup.team2.score || 0);
                            const margin = Math.abs(score1 - score2);
                            if (margin > stats.largest_margin.value) {
                                stats.largest_margin = {
                                    value: parseFloat(margin.toFixed(2)),
                                    winner: score1 > score2 ? matchup.team1.name : matchup.team2.name,
                                    loser: score1 > score2 ? matchup.team2.name : matchup.team1.name,
                                    week: weekNum,
                                    season: season
                                };
                            }
                        }
                    });
                }
            });
        }

        // Check Most Points in a Season
        Object.entries(seasonPoints).forEach(([team, points]) => {
            if (points > stats.most_points_season.value) {
                stats.most_points_season = { value: parseFloat(points.toFixed(2)), team: team, season: season };
            }
        });
    }

    // Round total points
    stats.total_points = parseFloat(stats.total_points.toFixed(2));

    try {
        await db.collection('stats').doc('all_time').set(stats);
        console.log('Successfully uploaded all-time stats.');
    } catch (error) {
        console.error('Error uploading stats:', error);
    }
}

async function main() {
    try {
        await uploadCollection('data/draft', 'draft');
        await uploadCollection('data/fantasy', 'fantasy');
        await calculateAndUploadStats();
        console.log('Data upload complete.');
    } catch (error) {
        console.error('Data upload failed:', error);
        process.exit(1);
    }
}

main();

File: scripts/upload_data.py
import firebase_admin
from firebase_admin import credentials
from firebase_admin import db
import json
import os
import sys

# Configuration
# 1. Download your service account key from Project Settings > Service Accounts
# 2. Rename it to 'serviceAccountKey.json' and place it in the project root
KEY_FILE = 'serviceAccountKey.json'
DATABASE_URL = "https://topina-9cd75-default-rtdb.firebaseio.com/"

def init_firebase():
    if not os.path.exists(KEY_FILE):
        print(f"Error: {KEY_FILE} not found.")
        print("Please download your Service Account Key from Firebase Console:")
        print("Project Settings -> Service Accounts -> Generate New Private Key")
        print(f"Save it as '{KEY_FILE}' in this directory.")
        return False

    cred = credentials.Certificate(KEY_FILE)
    firebase_admin.initialize_app(cred, {
        'databaseURL': DATABASE_URL
    })
    return True

def upload_collection(directory, node_name):
    dir_path = os.path.join(os.getcwd(), directory)
    
    if not os.path.exists(dir_path):
        print(f"Directory {dir_path} does not exist. Skipping.")
        return

    files = [f for f in os.listdir(dir_path) if f.endswith('.json')]
    print(f"Uploading {len(files)} files from {directory} to node '{node_name}'...")

    ref = db.reference(node_name)

    for filename in files:
        file_path = os.path.join(dir_path, filename)
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        # Use filename without extension as child key
        child_key = os.path.splitext(filename)[0]
        
        try:
            ref.child(child_key).set(data)
            print(f"✓ Uploaded {child_key}")
        except Exception as e:
            print(f"✗ Error uploading {child_key}: {e}")

def calculate_and_upload_stats():
    print("Calculating all-time stats...")
    fantasy_dir = os.path.join(os.getcwd(), 'data', 'fantasy')
    if not os.path.exists(fantasy_dir):
        return

    files = [f for f in os.listdir(fantasy_dir) if f.endswith('.json')]

    stats = {
        'seasons_count': len(files),
        'total_games': 0,
        'total_points': 0,
        'highest_score': {'value': 0, 'team': '', 'week': '', 'season': ''},
        'lowest_score': {'value': 1000, 'team': '', 'week': '', 'season': ''},
        'largest_margin': {'value': 0, 'winner': '', 'loser': '', 'week': '', 'season': ''},
        'most_points_season': {'value': 0, 'team': '', 'season': ''}
    }

    for filename in files:
        season = filename.split('_')[2].split('.')[0] # fantasy_data_2024.json -> 2024
        with open(os.path.join(fantasy_dir, filename), 'r', encoding='utf-8') as f:
            content = json.load(f)

        season_points = {}

        if 'weeks' in content:
            for week_num, week_data in content['weeks'].items():
                if 'matchups' in week_data:
                    for matchup in week_data['matchups']:
                        stats['total_games'] += 1
                        
                        # Helper to process team
                        def process_team(team_obj, opponent_obj=None):
                            if not team_obj: return
                            
                            score = float(team_obj.get('score', 0))
                            stats['total_points'] += score
                            
                            # High Score
                            if score > stats['highest_score']['value']:
                                stats['highest_score'] = {'value': score, 'team': team_obj['name'], 'week': week_num, 'season': season}
                            
                            # Low Score (ignore 0)
                            if score > 0 and score < stats['lowest_score']['value']:
                                stats['lowest_score'] = {'value': score, 'team': team_obj['name'], 'week': week_num, 'season': season}
                                
                            # Season Points
                            name = team_obj['name']
                            season_points[name] = season_points.get(name, 0) + score

                        process_team(matchup.get('team1'))
                        process_team(matchup.get('team2'))

                        # Margin
                        if matchup.get('team1') and matchup.get('team2'):
                            s1 = float(matchup['team1'].get('score', 0))
                            s2 = float(matchup['team2'].get('score', 0))
                            margin = abs(s1 - s2)
                            
                            if margin > stats['largest_margin']['value']:
                                winner = matchup['team1']['name'] if s1 > s2 else matchup['team2']['name']
                                loser = matchup['team2']['name'] if s1 > s2 else matchup['team1']['name']
                                stats['largest_margin'] = {
                                    'value': round(margin, 2),
                                    'winner': winner,
                                    'loser': loser,
                                    'week': week_num,
                                    'season': season
                                }

        # Most Points in Season
        for team, points in season_points.items():
            if points > stats['most_points_season']['value']:
                stats['most_points_season'] = {'value': round(points, 2), 'team': team, 'season': season}

    stats['total_points'] = round(stats['total_points'], 2)

    try:
        db.reference('stats/all_time').set(stats)
        print("✓ Uploaded all-time stats")
    except Exception as e:
        print(f"✗ Error uploading stats: {e}")

def main():
    if not init_firebase():
        sys.exit(1)
        
    upload_collection('data/draft', 'draft')
    upload_collection('data/fantasy', 'fantasy')
    calculate_and_upload_stats()
    print("Done!")

if __name__ == "__main__":
    main()
